File: benchmarks3d/many_pyramids3.rs
use rapier_testbed3d::Testbed;
use rapier3d::prelude::*;

fn create_pyramid(
    bodies: &mut RigidBodySet,
    colliders: &mut ColliderSet,
    offset: Vector<f32>,
    stack_height: usize,
    rad: f32,
) {
    let shift = rad * 2.0;

    for i in 0usize..stack_height {
        for j in i..stack_height {
            let fj = j as f32;
            let fi = i as f32;
            let x = (fi * shift / 2.0) + (fj - fi) * shift;
            let y = fi * shift;

            // Build the rigid body.
            let rigid_body = RigidBodyBuilder::dynamic().translation(vector![x, y, 0.0] + offset);
            let handle = bodies.insert(rigid_body);
            let collider = ColliderBuilder::cuboid(rad, rad, rad);
            colliders.insert_with_parent(collider, handle, bodies);
        }
    }
}

pub fn init_world(testbed: &mut Testbed) {
    /*
     * World
     */
    let mut bodies = RigidBodySet::new();
    let mut colliders = ColliderSet::new();
    let impulse_joints = ImpulseJointSet::new();
    let multibody_joints = MultibodyJointSet::new();

    let rad = 0.5;
    let pyramid_count = 40;
    let spacing = 4.0;

    /*
     * Ground
     */
    let ground_size = 50.0;
    let ground_height = 0.1;

    let rigid_body = RigidBodyBuilder::fixed().translation(vector![0.0, -ground_height, 0.0]);
    let ground_handle = bodies.insert(rigid_body);
    let collider = ColliderBuilder::cuboid(
        ground_size,
        ground_height,
        pyramid_count as f32 * spacing / 2.0 + ground_size,
    );
    colliders.insert_with_parent(collider, ground_handle, &mut bodies);

    /*
     * Create the cubes
     */
    for pyramid_index in 0..pyramid_count {
        let bottomy = rad;
        create_pyramid(
            &mut bodies,
            &mut colliders,
            vector![
                0.0,
                bottomy,
                (pyramid_index as f32 - pyramid_count as f32 / 2.0) * spacing
            ],
            20,
            rad,
        );
    }

    /*
     * Set up the testbed.
     */
    testbed.set_world(bodies, colliders, impulse_joints, multibody_joints);
    testbed.look_at(point![100.0, 100.0, 100.0], Point::origin());
}

File: bindings/python/rapier-testbed/rapier_testbed/examples3/stress_tests/many_pyramids3.py
"""Port of examples3d/stress_tests/many_pyramids3.rs."""
from __future__ import annotations

import rapier3d as rp
from ..._registry import register

CATEGORY = "Stress Tests"
NAME = "Many pyramids"


def _create_pyramid(bodies, colliders, offset, stack_height, rad):
    shift = rad * 2.0
    for i in range(stack_height):
        for j in range(i, stack_height):
            fj = float(j)
            fi = float(i)
            x = (fi * shift / 2.0) + (fj - fi) * shift
            y = fi * shift
            body = rp.RigidBody.dynamic().translation(
                (x + offset[0], y + offset[1], 0.0 + offset[2])
            )
            h = bodies.insert(body)
            colliders.insert_with_parent(rp.Collider.cuboid(rad, rad, rad), h, bodies)


def init_world(testbed) -> None:
    bodies = rp.RigidBodySet()
    colliders = rp.ColliderSet()
    impulse_joints = rp.ImpulseJointSet()
    multibody_joints = rp.MultibodyJointSet()

    rad = 0.5
    pyramid_count = 40
    spacing = 4.0

    ground_size = 50.0
    ground_height = 0.1
    ground = rp.RigidBody.fixed().translation((0.0, -ground_height, 0.0))
    ground_h = bodies.insert(ground)
    colliders.insert_with_parent(
        rp.Collider.cuboid(
            ground_size, ground_height, pyramid_count * spacing / 2.0 + ground_size
        ),
        ground_h,
        bodies,
    )

    for p in range(pyramid_count):
        bottomy = rad
        _create_pyramid(
            bodies,
            colliders,
            (0.0, bottomy, (p - pyramid_count / 2.0) * spacing),
            20,
            rad,
        )

    testbed.set_world(bodies, colliders, impulse_joints, multibody_joints)
    testbed.look_at((100.0, 100.0, 100.0), (0.0, 0.0, 0.0))


register(CATEGORY, NAME, init_world, dim=3)


if __name__ == "__main__":
    from ..._picker import run

    run(initial=f"{CATEGORY} / {NAME}")
